### PR TITLE
adding pitr support

### PR DESCRIPTION
**Pull request by mayanktiwari09**
Modified the following existing command:
cf start-restore SERVICE_INSTANCE_NAME BACKUP_GUID
New Commands Introduced:
cf start-restore SERVICE_INSTANCE_NAME --backup_guid BACKUP_GUID
cf start-restore SERVICE_INSTANCE_NAME --timestamp TIME_STAMP

### CODE REVIEW

File: service-fabrik-plugin.go [hunk 1]
```diff
@@ -261,7 +273,7 @@ func (serviceFabrikPlugin *ServiceFabrikPlugin) GetMetadata() plugin.PluginMetad
 				Name:     "start-restore",
 				HelpText: "Start restore of a service instance",
 				UsageDetails: plugin.Usage{
-					Usage: "cf start-restore SERVICE_INSTANCE_NAME BACKUP_ID",
+					Usage: "cf start-restore SERVICE_INSTANCE_NAME --backup_guid BACKUP_ID \n     cf start-restore SERVICE_INSTANCE_NAME --time_stamp TIME_STAMP",
```
Comment (abh1kg) [inline service-fabrik-plugin.go]: `timestamp` is a single word... No need to hyphenate

File: service-fabrik-plugin.go [hunk 2]
```diff
@@ -146,18 +146,30 @@ func (serviceFabrikPlugin *ServiceFabrikPlugin) Run(cliConnection plugin.CliConn
...
+				if args[2] == "--backup_guid" {
+					fmt.Println("Are you sure you want to start restore? (y/n)")
+					var userChoice string
+					fmt.Scanln(&userChoice)
+					if userChoice == "y" {
+						restore.NewRestoreCommand(cliConnection).StartRestore(cliConnection, args[1], args[3], "", true)
+					} else {
+						os.Exit(7)
+					}
+				} else if args[2] == "--time_stamp" {
```
Comment (abh1kg) [inline service-fabrik-plugin.go]: Change to `timestamp`

**Review by abh1kg — CHANGES_REQUESTED** [review]
Will check the code more stringently later but initial comments are provided

File: restore/restore.go [hunk 3]
```diff
@@ -96,18 +96,31 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
 	return resp
 }
 
-func (c *RestoreCommand) StartRestore(cliConnection plugin.CliConnection, serviceInstanceName string, backupId string) {
+func (c *RestoreCommand) StartRestore(cliConnection plugin.CliConnection, serviceInstanceName string, backupId string, timeStamp string, guidBool bool) {
```
Comment (pritishmishra) [inline restore/restore.go]: can we rename `guidBool` to `isGuidOperation` ?

File: restore/restore.go [hunk 4]
```diff
@@ -96,18 +96,31 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
...
-	var jsonprep string = `{"backup_guid": "` + backupId + `"}`
-	var jsonStr = []byte(jsonprep)
-	var req_body = bytes.NewBuffer(jsonStr)
+	var req_body = bytes.NewBuffer([]byte(""))
+	if guidBool == true {
+		var jsonprep string = `{"backup_guid": "` + backupId + `"}`
+		var jsonStr = []byte(jsonprep)
+		req_body = bytes.NewBuffer(jsonStr)
+	} else {
+		time, err := time.Parse(time.RFC3339, timeStamp)
```
Comment (pritishmishra) [inline restore/restore.go]: Add a comment here referring to what is RFC3339, may be a link or something for future reference.

Comment (pritishmishra) [inline restore/restore.go]: Avoid using the same variable name as the library name. The library name is time. So, use variable like, `parsedTimestamp`

File: restore/restore.go [hunk 5]
```diff
@@ -96,18 +96,31 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
...
+		if err != nil {
+			fmt.Println(AddColor("FAILED", red))
+			fmt.Println(err)
+			fmt.Println("Please enter time in ISO8061 format, example - 2018-11-12T11:45:26.371Z")
```
Comment (pritishmishra) [inline restore/restore.go]: If you support additional formats, can you give an example of them here too?
May be one or two more? Just to help the user understand his options.
e.g. `2018-11-12T11:45`

File: restore/restore.go [hunk 6]
```diff
@@ -96,18 +96,31 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
...
+			return
+		}
+		var jsonprep string = `{"time_stamp": "` + time.String() + `"}`
```
Comment (pritishmishra) [inline restore/restore.go]: change `time_stamp` to `timestamp` and time to something else as explained above.

**Review by pritishmishra — CHANGES_REQUESTED** [review]
Overall looks perfect!
Just a few minor comments related to variable names.

File: restore/restore.go (hunk 6 above)
Comment (pritishmishra) [reply restore/restore.go]: ok.. may be here you need `time_stamp`, because the API expects it? Is that the case here?

File: restore/restore.go [hunk 7]
```diff
@@ -134,7 +149,11 @@ func (c *RestoreCommand) StartRestore(cliConnection plugin.CliConnection, servic
```
Comment (sauravmndl-zz) [inline restore/restore.go:149]: Please check the error message in case of error.

File: restore/restore.go [hunk 8]
```diff
@@ -96,18 +97,32 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
...
 
 	client := GetHttpClient()
-
-	var jsonprep string = `{"backup_guid": "` + backupId + `"}`
-	var jsonStr = []byte(jsonprep)
-	var req_body = bytes.NewBuffer(jsonStr)
+	var req_body = bytes.NewBuffer([]byte(""))
+	if guidBool == true {
+		var jsonprep string = `{"backup_guid": "` + backupId + `"}`
+		var jsonStr = []byte(jsonprep)
```
Comment (sauravmndl-zz) [inline restore/restore.go]: If you have **userSpaceGuid** with you please send in body as that would cut down one call at broker end.

File: restore/restore.go (hunk 5 above)
Comment (sauravmndl-zz) [reply restore/restore.go]: Format should be ISO8601 .

File: restore/restore.go [hunk 9]
```diff
@@ -96,18 +97,32 @@ func GetResponse(client *http.Client, req *http.Request) *http.Response {
...
+	} else {
+		time, err := time.Parse(time.RFC3339, timeStamp)
+		if err != nil {
+			fmt.Println(AddColor("FAILED", red))
+			fmt.Println(err)
+			fmt.Println("Please enter time in ISO8061 format, example - 2018-11-12T11:45:26.371Z")
+			return
+		}
+		var epochTime string = strconv.FormatInt(time.UnixNano()/1000000, 10)
+                var jsonprep string = `{"time_stamp": "`+ epochTime + `"}`
```
Comment (sauravmndl-zz) [inline restore/restore.go]: Follow camel case as followed everywhere jsonprep-->jsonPrep

**Review by sauravmndl-zz — CHANGES_REQUESTED** [review]

**Review by sauravmndl-zz — APPROVED** [review]

**Comment by kamath-prasad** [comment]
@mayanktiwari09 : can this be merged ?

@pritishmishra  / @abh1kg : can you approve ?

**Review by pritishmishra — APPROVED** [review]
Good to go!

**Review by abh1kg — APPROVED** [review]

**Pull request merged**